torch/optim/lr_scheduler.py

```diff
@@ -0,0 +1,206 @@
+
+
+class ReduceLROnPlateau(object):
+    """Reduce learning rate when a metric has stopped improving.
+    Models often benefit from reducing the learning rate by a factor
+    of 2-10 once learning stagnates. This scheduler reads a metrics
+    quantity and if no improvement is seen for a 'patience' number
+    of epochs, the learning rate is reduced.
+
+    Args:
+        factor: factor by which the learning rate will
+            be reduced. new_lr = lr * factor
+        patience: number of epochs with no improvement
+            after which learning rate will be reduced.
+        verbose: int. 0: quiet, 1: update messages.
+        mode: one of {min, max}. In `min` mode,
+            lr will be reduced when the quantity
+            monitored has stopped decreasing; in `max`
+            mode it will be reduced when the quantity
+            monitored has stopped increasing.
+        threshold: threshold for measuring the new optimum,
+            to only focus on significant changes.
+        threshold_mode: one of {rel, abs}. In 'rel' mode,
+            dynamic_threshold = best * ( 1 + threshold )
+            in 'max' mode or best * ( 1 - threshold ) in
+            'min' mode. In 'abs' mode, dynamic_threshold =
+            best + threshold in 'max' mode or
+            best - threshold in 'min' mode.
+        cooldown: number of epochs to wait before resuming
+            normal operation after lr has been reduced.
+        min_lr: lower bound on the learning rate.
+
+
+    Example:
+        >>> optimizer = torch.optim.SGD(model.parameters(), lr=0.1, momentum=0.9)
+        >>> scheduler = ReduceLROnPlateau(optimizer, 'min')
+        >>> for epoch in range(10):
+        >>>     train(...)
+        >>>     val_loss = validate(...)
+        >>>     # different from LambdaLR, step should be called after validate()
+        >>>     scheduler.step(epoch, val_loss)
+    """
+
+    def __init__(self, optimizer, mode='min', factor=0.1, patience=10,
+                 verbose=0, threshold=1e-4, threshold_mode='rel',
+                 cooldown=0, min_lr=0):
+
+        if factor >= 1.0:
+            raise ValueError('ReduceLROnPlateau '
+                             'does not support a factor >= 1.0.')
+        self.factor = factor
+        self.min_lr = min_lr
+        self.patience = patience
+        self.verbose = verbose
+        self.cooldown = cooldown
+        self.cooldown_counter = 0  # Cooldown counter.
+        self.monitor_op = _MonitorOp(mode=mode, threshold=threshold,
+                                     threshold_mode=threshold_mode)
+        self.wait = 0
+        self.best = 0
+        assert isinstance(optimizer, Optimizer)
+        self.optimizer = optimizer
+        self._reset()
+
+    def _reset(self):
+        """Resets wait counter and cooldown counter.
+        """
+        self.best = self.monitor_op.worse
+        self.cooldown_counter = 0
+        self.wait = 0
+        self.lr_epsilon = self.min_lr * 1e-4
```

Comment:
Agree with the init part. This should be related to min lr. Otherwise the code would be wrong when the user sets a very low lr.